source/jormungandr/jormungandr/scenarios/experimental.py

```diff
@@ -595,172 +51,97 @@ def call_kraken(self, request_type, request, instance, krakens_call):
-                j.internal_id = str(generate_id())
-                response_tuples.append((dep_mode, arr_mode, j))
-            resp.append(local_resp)
-
-        # Once the pt journey is found, we need to reconstruct the whole journey with fallback regarding the mode
-
-        # For the sake of performance, we compute at first all fallback direct path asynchronously
-        # then we update the pool of direct paths
-        futures = worker.get_fallback_direct_path_futures(response_tuples, crowfly_stop_points, odt_stop_points)
-        for future in gevent.iwait(futures):
-            resp_key, resp_direct_path = future.get()
-            g.fallback_direct_path_pool[resp_key] = resp_direct_path
-
-        # Now we construct the whole journey by concatenating the fallback direct path with the pt journey
-        worker.build_journeys(response_tuples, crowfly_stop_points, odt_stop_points)
-
-        #If resp doesn't contain any response we have to add an error message
-        if len(resp) == 0:
-            if len(g.origins_fallback[dep_mode]) == 0 and len(g.destinations_fallback[arr_mode]) == 0:
-                resp.append(self._make_error_response("no origin point nor destination point",
-                                                      response_pb2.Error.no_origin_nor_destination))
-            elif len(g.origins_fallback[dep_mode]) == 0:
-                resp.append(self._make_error_response("no origin point", response_pb2.Error.no_origin))
-            elif len(g.destinations_fallback[arr_mode]) == 0:
-                resp.append(self._make_error_response("no destination point", response_pb2.Error.no_destination))
-            return resp
-        for r in resp:
+        #
+        for mode in requested_dep_modes:
+            direct_path_pool.add_async_request(requested_orig_obj, requested_dest_obj, mode,
+                                               fallback_extremity, request, direct_path_type)
+
+        # if max_duration(time to pass in pt) is zero, there is no need to continue,
+        # we return all direct path without pt
+        if not request.get('max_duration', 0):
+            res = []
+            for mode in requested_dep_modes:
+                dp = direct_path_pool.wait_and_get(requested_orig_obj, requested_dest_obj, mode,
+                                                   fallback_extremity, direct_path_type)
+                res.append(dp)
+            return res
+
+        orig_proximities_by_crowfly = ProximitiesByCrowflyPool(instance, requested_orig_obj, requested_dep_modes,
+                                                               request, direct_path_pool)
+        dest_proximities_by_crowfly = ProximitiesByCrowflyPool(instance, requested_dest_obj, requested_arr_modes,
+                                                               request, None)
+
+        is_origin= True
+        orig_fallback_durations_pool = FallbackDurationsPool(instance, requested_orig_obj, is_origin, requested_dep_modes,
```

Comment:
for those function with lots of parameters I think it's better to name them all (and this way you can remove the `is_origin` variable